src/v/datalake/translation/scheduling.h

```diff
@@ -0,0 +1,333 @@
+     */
+    virtual void start_translation(clock::duration time_slice) = 0;
+
+    /**
+     * Invoked by the translator to preempt an inprogress translation. All
+     * resources should be released and the translator should invoke \ref
+     * scheduling_notifications::notify_done when done.
+     */
+    virtual void stop_translation() = 0;
+};
+
+std::ostream& operator<<(std::ostream&, const translator&);
+
+struct scheduler_state;
+class scheduler;
+
+/**
+ * Encapsulates all the scheduling state for a single translator.
+ *   ┌───────┐
+ *   │started│
+ *   └┬──────┘
+ *   ┌▽────────┐
+ *   │  idle   │
+ *   └△───────┬┘
+ *   ┌┴──────┐│
+ *   │running││ (notify_ready())
+ *   └△──────┘│
+ *   ┌┴───────▽─┐
+ *   │ waiting  │
+ *   └──────────┘
+ *
+ * Summary of state transitions
+ * * idle -> waiting  on notify_ready()
+ * * waiting -> running after scheduling
+ * * running -> idle on notify_done()
+ */
+
+class translator_state {
+public:
+    explicit translator_state(std::unique_ptr<translator> translator)
+      : _translator(std::move(translator)) {}
+    translator_state(const translator_state&) = delete;
+    translator_state& operator=(const translator_state&) = delete;
+    translator_state(translator_state&& other) noexcept;
+    translator_state& operator=(translator_state&&) noexcept;
+    ~translator_state() = default;
+
+    bool stop_in_progress() const { return _stop_in_progress; }
+
+    std::optional<clock::time_point> current_translation_start_time() const {
+        return _current_translation_start_time;
+    }
+
+    // hook into the waiting queue
+    intrusive_list_hook _waiting_hook;
+    // hook into the running queue
+    intrusive_list_hook _running_hook;
+
+    friend std::ostream& operator<<(std::ostream&, const translator_state&);
+
+private:
+    friend struct scheduler_state;
+    friend class scheduler;
+    std::unique_ptr<translator> _translator;
+    // Set for the inprogress translation.
+    std::optional<clock::time_point> _current_translation_start_time;
+    bool _stop_in_progress{false};
+};
+
+using translators = chunked_hash_map<translator_id, translator_state>;
+
+/**
+ * Encapsulates all the (internal) state the scheduler works with.
+ *
+ * Note to the users of class members. Be mindful of scheduling points
+ * when iterating through these containers. There is no explicit locking
+ * in place to avoid racy access across scheduling points.
+
+ * running/waiting lists
+ *
+ * General guideline is to not iterate thorough these lists across scheduling
+ * points. These lists are typically small ands its ok to loop through them in
+ * a single continuation.
+
+ * translators map
+ *
+ * For the most part there is no reason to loop through this map as all
+ * operations are scoped to a particular translator instance.
+ */
+struct scheduler_state {
+    void start_translation(translator_state&, clock::duration time_slice);
+    void stop_translation(translator_state&);
```

Comment:
Might just be me, but reading through this I kept finding it slightly confusing to follow because we refer to a couple things as `state` and it's easy to conflate <x>_state for <x>. Taking a step back, what do you think about the following renames:
- scheduler_state -> executor: tracks the registered translations and carries out the _execution_ of translation/stop requests
- translation_state -> translation_executable: a hook into the executor, as indicated by the "execut-" part of the name

Then the differences between scheduler and scheduler_state become clearer IMO: the scheduler repeatedly evaluates the policy and listens for signals in the system, while the scheduler_state/executor just perform tasks at the request of the scheduler.

Reply:
done, renamed.